lib/krane/kubernetes_resource.rb

```diff
@@ -44,38 +44,70 @@ class KubernetesResource
     SYNC_DEPENDENCIES = []
 
     class << self
-      def build(namespace: nil, context:, definition:, logger:, statsd_tags:, crd: nil, global_names: [])
+      def build(namespace: nil, context:, definition:, logger:, statsd_tags:, crd: nil, group_kinds: [])
         validate_definition_essentials(definition)
+
+        group = ::Krane::KubernetesResource.group_from_api_version(definition["apiVersion"])
+        kind = definition["kind"]
+        group_kind = ::Krane::KubernetesResource.combine_group_kind(group, kind)
+
+        klass = ::Krane::KubernetesResource.group_kind_to_const(group_kind)
+
         opts = { namespace: namespace, context: context, definition: definition, logger: logger,
                  statsd_tags: statsd_tags }
-        if (klass = class_for_kind(definition["kind"]))
-          return klass.new(**opts)
-        end
-        if crd
+
+        inst = if klass
+          klass.new(**opts)
+        elsif crd
           CustomResource.new(crd: crd, **opts)
         else
-          type = definition["kind"]
-          inst = new(**opts)
-          inst.type = type
-          inst.global = global_names.map(&:downcase).include?(type.downcase)
-          inst
+          new(**opts)
         end
-      end
 
-      def class_for_kind(kind)
-        if Krane.const_defined?(kind, false)
-          Krane.const_get(kind, false)
+        inst.global = GLOBAL
+        if (entry = group_kinds.find { |x| x["group_kind"] == group_kind })
+          inst.global = !entry["namespaced"]
         end
-      rescue NameError
-        nil
+        inst
       end
 
       def timeout
         self::TIMEOUT
       end
 
+      def group
```

Comment:
The only edge case I see for `::group` and `::kind` is if we call it on `CustomResource`, in which case the result wouldn't make sense. I don't think that's going to be a problem in practice, though

Reply:
Why doesn't it make sense for the CR?

Reply:
Because this method is generating its result based on _class name_ within `krane`. For custom resources this ends up being: `group(cr) == ""` and `kind(cr) == CustomResource`, but this isn't reflective of any actual object on the API-server.

Ultimately I don't think it's a big deal, but just wanted to point it out in case this violated an assumption somewhere

Reply:
This only happens when you call it on the class, not on an actual instance of CustomResource.